cmd/erasure-multipart.go

```diff
@@ -987,6 +950,80 @@ func (er erasureObjects) ListObjectParts(ctx context.Context, bucket, object, up
 	return result, nil
 }
 
+func readParts(ctx context.Context, disks []StorageAPI, bucket string, partMetaPaths []string, partNumbers []int, readQuorum int) ([]*ObjectPartInfo, error) {
+	g := errgroup.WithNErrs(len(disks))
+
+	objectPartInfos := make([][]*ObjectPartInfo, len(disks))
+	// Rename file on all underlying storage disks.
+	for index := range disks {
+		index := index
+		g.Go(func() (err error) {
+			if disks[index] == nil {
+				return errDiskNotFound
+			}
+			objectPartInfos[index], err = disks[index].ReadParts(ctx, bucket, partMetaPaths...)
+			return err
+		}, index)
+	}
+
+	if err := reduceReadQuorumErrs(ctx, g.Wait(), objectOpIgnoredErrs, readQuorum); err != nil {
+		return nil, err
+	}
+
+	partInfosInQuorum := make([]*ObjectPartInfo, len(partMetaPaths))
+	partMetaQuorumMap := make(map[string]int, len(partNumbers))
+	for pidx := range partMetaPaths {
+		var pinfo *ObjectPartInfo
+		for idx := range disks {
+			if objectPartInfos[idx] == nil || len(objectPartInfos[idx]) == 0 {
+				continue
+			}
+
+			pinfo = objectPartInfos[idx][pidx]
+			if pinfo == nil {
+				continue
+			}
+
+			if pinfo.ETag != "" {
+				partMetaQuorumMap[pinfo.ETag]++
+			} else {
+				partMetaQuorumMap[partMetaPaths[pidx]]++
+			}
+		}
+		if pinfo == nil {
+			partInfosInQuorum[pidx] = &ObjectPartInfo{
+				Error: InvalidPart{
+					PartNumber: partNumbers[pidx],
+				}.Error(),
+			}
+			continue
+		}
+		if pinfo.ETag != "" && partMetaQuorumMap[pinfo.ETag] >= readQuorum {
```

Comment:
IIUC, we can't rely on `pinfo` since it holds the last disk's part information. You need to range over the two maps which maintain counts indexed by etag and path names.